<a href="https://colab.research.google.com/github/LilInformat/UofT-3253-OzoneAnalysis/blob/master/OzoneNotebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Forecasting Ozone Days Pt.1

In [1]:
!pip install openml

     |████████████████████████████████| 163kB 9.8MB/s 
  Created wheel for openml: filename=openml-0.10.2-cp36-none-any.whl size=190318 sha256=620e30249f56cc49ebf9a5f748ecd3f5f663865d5463c56ca9218fcf52cccac3
  Stored in directory: /root/.cache/pip/wheels/71/ec/5f/aaad9e184680b0b8f1a02ff0ec640cace5adf5bff7bb0af1b4
  Created wheel for liac-arff: filename=liac_arff-2.4.0-cp36-none-any.whl size=13333 sha256=102f953445e6e9ad6f98d4f20d388ac56033bb9019af9753e6aff0ecc564c06e
  Stored in directory: /root/.cache/pip/wheels/d1/6a/e7/529dc54d76ecede4346164a09ae3168df358945612710f5203
Successfully built openml liac-arff


In [0]:
import openml
import os
import pandas as pd
import numpy as np
from tqdm import tqdm

from sklearn.metrics import accuracy_score, roc_auc_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.model_selection import train_test_split, cross_validate, StratifiedKFold, GridSearchCV, cross_val_score
from sklearn.svm import SVC, LinearSVC
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

pd.options.display.max_rows = 4000

openml.config.cache_directory = os.path.expanduser('/')
# openml.config.apikey = 'YOURKEY'

## Understand

Here we present the problem in which we wish to explore in this notebook, in that we wish to be able to forecast days in which Ozone levels are high. In this notebook, we will be looking at the 8hr peak dataset whose data was collected from 1998 to 2004 at the Houston, Galveston and Brazoria area.

A list of the features that are provided in this dataset can be found here:

https://archive.ics.uci.edu/ml/machine-learning-databases/ozone/

In total there are 72 Features and a single Target (DayClass). Using the features provided we will try to train a model that can forecast if it is an Ozone day or not.

## Load

In [18]:
# Iris dataset https://www.openml.org/d/1487
dataset = openml.datasets.get_dataset(1487)

# Print a summary
print(f"This is dataset '{dataset.name}', the target feature is "
      f"'{dataset.default_target_attribute}'")
print(f"URL: {dataset.url}")

This is dataset 'ozone-level-8hr', the target feature is 'Class'
URL: https://www.openml.org/data/v1/download/1592279/ozone-level-8hr.arff


In [0]:
raw_data = dataset.get_data(dataset_format='dataframe')

In [0]:
#Column Names can be found here: https://archive.ics.uci.edu/ml/machine-learning-databases/ozone/
df = raw_data[0]
df.columns= ['WSR0','WSR1','WSR2','WSR3','WSR4','WSR5','WSR6','WSR7','WSR8','WSR9','WSR10','WSR11','WSR12',
             'WSR13','WSR14','WSR15','WSR16','WSR17','WSR18','WSR19','WSR20','WSR21','WSR22','WSR23','WSR_PK',
             'WSR_AV','T0','T1','T2','T3','T4','T5','T6','T7','T8','T9','T10','T11','T12','T13','T14','T15',
             'T16','T17','T18','T19','T20','T21','T22','T23','T_PK','T_AV','T85','RH85','U85','V85','HT85','T70',
             'RH70','U70','V70','HT70','T50','RH50','U50','V50','HT50','KI','TT','SLP','SLP_','Precp','DayClass']

## Explore

In [23]:
print(df.shape)
df.head(10)

(2534, 73)


WSR0  WSR1  WSR2  WSR3  ...           SLP       SLP_  Precp  DayClass
0   0.8   1.8   2.4   2.1  ...  10330.000000 -55.000000   0.00         1
1   2.8   3.2   3.3   2.7  ...  10275.000000 -55.000000   0.00         1
2   2.9   2.8   2.6   2.1  ...  10235.000000 -40.000000   0.00         1
3   4.7   3.8   3.7   3.8  ...  10195.000000 -40.000000   2.08         1
4   2.6   2.1   1.6   1.4  ...  10164.198442  -0.119949   0.58         1
5   3.1   3.5   3.3   2.5  ...  10120.000000  -0.119949   5.84         1
6   3.7   3.2   3.8   5.1  ...  10040.000000 -80.000000   0.18         1
7   2.2   2.9   3.4   4.2  ...  10065.000000  25.000000   0.00         1
8   1.0   1.5   1.2   1.2  ...  10120.000000  55.000000   0.00         1
9   0.9   0.6   0.5   0.5  ...  10180.000000  60.000000   0.00         1

[10 rows x 73 columns]

In [22]:
df.describe()

WSR0         WSR1  ...         SLP_        Precp
count  2534.000000  2534.000000  ...  2534.000000  2534.000000
mean      1.640179     1.586351  ...    -0.119949     0.372026
std       1.194568     1.191320  ...    34.693349     1.317438
min       0.000000     0.000000  ...  -135.000000     0.000000
25%       0.800000     0.700000  ...   -20.000000     0.000000
50%       1.600000     1.500000  ...    -0.119949     0.000000
75%       2.200000     2.100000  ...    15.000000     0.050000
max       7.500000     7.700000  ...   140.000000    20.650000

[8 rows x 72 columns]

We only have 2534 rows of data. Additionally, the values of these features are decimal and can be negative as well as be very large. This implies that we may want to use a MinMaxScaler depending on which model we train the data on. 

Notice that the describe did not show the DayClass column in the data.

In [30]:
print(df['DayClass'].values.unique())
print(df['DayClass'].dtype)

[1, 2]
Categories (2, object): [1 < 2]
category


We may need to convert this output to a numerical value as some classifiers may expect this value to be numerical. Since there are only two values, we can set 1's to 0 and 2's to 1.

In [32]:
print(df.isna().sum())

WSR0        0
WSR1        0
WSR2        0
WSR3        0
WSR4        0
WSR5        0
WSR6        0
WSR7        0
WSR8        0
WSR9        0
WSR10       0
WSR11       0
WSR12       0
WSR13       0
WSR14       0
WSR15       0
WSR16       0
WSR17       0
WSR18       0
WSR19       0
WSR20       0
WSR21       0
WSR22       0
WSR23       0
WSR_PK      0
WSR_AV      0
T0          0
T1          0
T2          0
T3          0
T4          0
T5          0
T6          0
T7          0
T8          0
T9          0
T10         0
T11         0
T12         0
T13         0
T14         0
T15         0
T16         0
T17         0
T18         0
T19         0
T20         0
T21         0
T22         0
T23         0
T_PK        0
T_AV        0
T85         0
RH85        0
U85         0
V85         0
HT85        0
T70         0
RH70        0
U70         0
V70         0
HT70        0
T50         0
RH50        0
U50         0
V50         0
HT50        0
KI          0
TT          0
SLP         0
SLP_        0
Precp 

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2534 entries, 0 to 2533
Data columns (total 73 columns):
WSR0        2534 non-null float64
WSR1        2534 non-null float64
WSR2        2534 non-null float64
WSR3        2534 non-null float64
WSR4        2534 non-null float64
WSR5        2534 non-null float64
WSR6        2534 non-null float64
WSR7        2534 non-null float64
WSR8        2534 non-null float64
WSR9        2534 non-null float64
WSR10       2534 non-null float64
WSR11       2534 non-null float64
WSR12       2534 non-null float64
WSR13       2534 non-null float64
WSR14       2534 non-null float64
WSR15       2534 non-null float64
WSR16       2534 non-null float64
WSR17       2534 non-null float64
WSR18       2534 non-null float64
WSR19       2534 non-null float64
WSR20       2534 non-null float64
WSR21       2534 non-null float64
WSR22       2534 non-null float64
WSR23       2534 non-null float64
WSR_PK      2534 non-null float64
WSR_AV      2534 non-null float64
T0       

In [0]:
target_attr = 'DayClass'
target_vals = {'ozone_day': 2, 'normal_day': 1}

In [10]:
X = df.iloc[:, 1:-1]
y = df[target_attr]
print('X:', X.shape)
print('y:', y.shape)

X: (2534, 71)
y: (2534,)


In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, shuffle=True)

In [0]:
scaler = StandardScaler()
svm = LinearSVC(C=1, loss="hinge", random_state=42)

In [13]:
scaled_svm = Pipeline([
        ("scaler", scaler),
        ("linear_svc", svm),
])
scaled_svm.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


Pipeline(memory=None,
         steps=[('scaler',
                 StandardScaler(copy=True, with_mean=True, with_std=True)),
                ('linear_svc',
                 LinearSVC(C=1, class_weight=None, dual=True,
                           fit_intercept=True, intercept_scaling=1,
                           loss='hinge', max_iter=1000, multi_class='ovr',
                           penalty='l2', random_state=42, tol=0.0001,
                           verbose=0))],
         verbose=False)

In [0]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())

In [15]:
svm_scores = cross_val_score(SVC(),
                             X_train,
                             y_train,
                             scoring="neg_mean_squared_error",
                             cv=10)
svm_rmse_scores = np.sqrt(-svm_scores)
display_scores(svm_rmse_scores)

/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:193: FutureWarning: The default 

Scores: [0.2506536  0.24061325 0.24061325 0.24061325 0.24061325 0.24061325
 0.24061325 0.24061325 0.24061325 0.24124895]
Mean: 0.24168085634483472
